applications: nrf_desktop: Update HID provider consumer and system control doc (#24641)
nrfconnect/sdk-nrf: applications/nrf_desktop/doc/hid_provider_system_ctrl.rst

.. _nrf_desktop_hid_provider_system_ctrl:

HID provider system control module
####################################

.. contents::
   :local:
   :depth: 2

> The HID provider system control module is a HID report provider integrated with :ref:`nrf_desktop_hid_state`.

Reviewer: ```suggestion
The HID provider system control module is a HID report provider integrated with the :ref:`nrf_desktop_hid_state`.
```

Configuration
*************

You can enable the default implementation of the HID provider using the :ref:`CONFIG_DESKTOP_HID_REPORT_PROVIDER_SYSTEM_CTRL <config_desktop_app_options>` Kconfig option.
This option is enabled by default if the device uses HID provider events (:ref:`CONFIG_DESKTOP_HID_REPORT_PROVIDER_EVENT <config_desktop_app_options>`) and supports HID system control reports (:ref:`CONFIG_DESKTOP_HID_REPORT_SYSTEM_CTRL_SUPPORT <config_desktop_app_options>`).

Alternatively, you can substitute the module with a custom HID system control report provider implementation.
Enable the :ref:`CONFIG_DESKTOP_HID_REPORT_PROVIDER_SYSTEM_CTRL_ALT <config_desktop_app_options>` Kconfig option to use a custom HID system control report provider.
The option disables the default HID system control report provider.
Make sure to introduce the custom HID system control report provider if you enable this option.
See the :ref:`nrf_desktop_hid_state_providing_hid_input_reports` documentation section of the HID state module for details.

> Default implementation

Reviewer: We could use .. include:: for this section as the same is used in the hid_provider_consumer_ctrl.rst file - Default implementation.

======================

> The module relies on keypresses (:c:struct:`button_event`) as the only source of user input.

Reviewer: ```suggestion
The module relies on keypresses (:c:struct:`button_event`) as the only source of your input.
```

Author: https://github.com/nrfconnect/sdk-nrf/pull/24641#discussion_r2367982015

The module needs to perform the following tasks:

* Identify keypresses related to the handled HID input report.

> * Queue a sequence of the user's key presses and releases that happens before connecting to a HID host.

Reviewer: ```suggestion
* Queue a sequence of your key presses and releases that happens before connecting to a HID host.
```

Author: https://github.com/nrfconnect/sdk-nrf/pull/24641#discussion_r2367982015

The :ref:`nrf_desktop_keys_state` is used to track the state of active keys after the connection with the HID host is established.
The module selects the :ref:`CONFIG_DESKTOP_KEYS_STATE <config_desktop_app_options>` Kconfig option to enable the utility.

Implementation details
**********************

On initialization, the module announces its presence by sending the :c:struct:`hid_report_provider_event` event with an appropriate report ID and implementation of callbacks from HID report provider API (:c:struct:`hid_report_provider_api`).
The :ref:`nrf_desktop_hid_state` receives the event and fills it with its own implementation of callbacks from the HID state API (:c:struct:`hid_state_api`).
After that process, the modules can communicate by callbacks.
The module also subscribes to the :c:struct:`button_event` event to get information about the button presses.
The module sends :c:struct:`hid_report_event` events to an appropriate subscriber when it is requested by the :ref:`nrf_desktop_hid_state`.

.. note::
    The HID report formatting function must work according to the HID report descriptor (``hid_report_desc``).

>     The source file containing the descriptor is provided by the :ref:`CONFIG_DESKTOP_HID_REPORT_DESC <config_desktop_app_options>` Kconfig option.

Author: See the comment above

Make sure to configure the HID keymap utility.
See the utility's documentation for details.

HID eventq
----------

> The :ref:`nrf_desktop_hid_eventq` is used to temporarily enqueue key state changes before connection with the HID host is established.

Reviewer: ```suggestion
You can use the :ref:`nrf_desktop_hid_eventq` to temporarily enqueue key state changes before connection with the HID host is established.
```

~~~~~~~~~~~~~~~~~

With the :ref:`CONFIG_DESKTOP_HID_REPORT_PROVIDER_SYSTEM_CTRL_KEYPRESS_EXPIRATION <config_desktop_app_options>` Kconfig option, you can set the amount of time after which a queued key will be considered expired.
For backwards compatibility, you can set the default value for this option using the deprecated :ref:`CONFIG_DESKTOP_HID_REPORT_EXPIRATION <config_desktop_app_options>` Kconfig option.
The higher the value, the longer the period from which the nRF Desktop application will recall pressed keys when the connection with HID host is established.

Keys state
----------

The :ref:`nrf_desktop_keys_state` is used to track the state of active keys after the connection with the HID host is established.
The module selects the :ref:`CONFIG_DESKTOP_KEYS_STATE <config_desktop_app_options>` Kconfig option to enable the utility.

Implementation details
**********************

The module is used by :ref:`nrf_desktop_hid_state` as a HID input report provider for the HID system control input report.
On initialization, the module submits the :c:struct:`hid_report_provider_event` event to establish two-way callbacks between the |hid_state| and the HID report provider.

> Handling keypresses

Reviewer: We could use .. include:: for this section as the same is used in the hid_provider_consumer_ctrl.rst file - Handling keypresses.
